This folder contains pre-built X-matrix and relevant files of species

#### How to generate supporting files of Scasa for new annotation can be [found here](https://github.com/eudoraleer/scasa/wiki/How-to-run-Scasa-for-a-new-annotation)

Currently, pre-built X-matrix and relevant files of some species are provided:
- Drosophila_melanogaster.BDGP6.32.108 from Ensembl: this folder (https://github.com/eudoraleer/scasa/tree/main/Anno)
- Homo_sapiens.GRCh38.106 from Ensembl can be downloaded [from here](https://www.dropbox.com/s/ppe4oqeoxv0jkb8/Homo_sapiens_GRCh38_106_Anno.zip).
- Homo_sapiens_GENCODE_42 version 42 can be downloaded [from here](https://www.dropbox.com/scl/fi/rb87wjbhaxoup7b3yqmrr/Homo_sapiens_GENCODE_42_Anno.zip).
  
##### It is noted that we have not tested the performance of scasa for those annotations

#### To run Scasa with new annotation
Use the clean version of fasta sequences **\*_clean.fa** provided in the folder of new annotation for downstream analysis. File *_clean.fa, which was generated by gen_tx2gene.R, keeps only transcripts existing in the corresponding gtf file. Then it is also required to update annotation files in scasa/REFERENCE/ by the files in the OUTPUT folder.

``` sh
#update txp2geneFile, newxmatrixFile and isoformMapperFile in the scasa/REFERENCE/
cp OUTPUT/X_matrix.RData $scasaPath/REFERENCE/Xmatrix_hg38_alevin.RData
cp OUTPUT/isoformMapper.RData $scasaPath/REFERENCE/isoform_groups_UCSC_hg38_alevin.RData
cp OUTPUT/txp2gene.tsv $scasaPath/REFERENCE/txp2gene_hg38.tsv

#now you can run scasa as the instruction on the scasa webpage: https://github.com/eudoraleer/scasa

```
